dynamodb-stream-reader/src/main/java/no/unit/nva/dynamodb/DynamoDBEventTransformer.java

```diff
@@ -40,90 +36,87 @@ public DynamoDBEventTransformer() {
-                break;
-            case CREATED_DATE_KEY:
-                document.setCreatedDate(value);
-                break;
-            case MODIFIED_DATE_KEY:
-                document.setModifiedDate(value);
-                break;
-            case DATE_YEAR:
-                document.setDate(value);
-                break;
-            default:
-                logger.debug(UNKNOWN_VALUE_KEY_MESSAGE, valueKey);
-                break;
-        }
-    }
-
-    private boolean isASimpleValue(AttributeValue attributeValue) {
-        return attributeValue.getM() == null && attributeValue.getL() == null;
-    }
-
-    private String addIndexPrefix(String prefix, String k) {
-        if (prefix != null && !prefix.isEmpty()) {
-            return prefix + Constants.SIMPLE_DOT_SEPARATOR + k;
-        } else {
-            return k;
-        }
+    public IndexDocument parseStreamRecord(DynamodbEvent.DynamodbStreamRecord streamRecord) {
+        JsonNode record = toJsonNode(streamRecord);
+
+        return new IndexDocument.Builder()
+                .withType(extractType(record))
+                .withIdentifier(extractIdentifier(record))
+                .withContributors(extractContributors(record))
+                .withDate(extractDate(record))
+                .withTitle(extractTitle(record))
+                .build();
+    }
+
+    private String extractDate(JsonNode record) {
+        return formatDate(extractYear(record), extractMonth(record), extractDay(record));
+    }
+
+    private String extractDay(JsonNode record) {
+        return textFromNode(record, DAY_JSON_POINTER);
+    }
+
+    private String extractMonth(JsonNode record) {
+        return textFromNode(record, MONTH_JSON_POINTER);
+    }
+
+    private String extractYear(JsonNode record) {
+        return textFromNode(record, YEAR_JSON_POINTER);
+    }
+
+    private List<IndexContributor> extractContributors(JsonNode record) {
+        return toStream(record.at(CONTRIBUTOR_LIST_JSON_POINTER))
+                .map(this::extractIndexContributor)
+                .collect(Collectors.toList());
+    }
+
+    private IndexContributor extractIndexContributor(JsonNode jsonNode) {
+        String identifier = textFromNode(jsonNode, CONTRIBUTOR_ARP_ID_JSON_POINTER);
+        String name = textFromNode(jsonNode, CONTRIBUTOR_NAME_JSON_POINTER);
+        return nonNull(name) ? generateIndexContributor(identifier, name) : null;
+    }
+
+    private String extractIdentifier(JsonNode record) {
+        return textFromNode(record, IDENTIFIER_JSON_POINTER);
+    }
+
+    private String extractTitle(JsonNode record) {
+        return textFromNode(record, MAIN_TITLE_JSON_POINTER);
+    }
+
+    private String extractType(JsonNode record) {
+        return textFromNode(record, TYPE_JSON_POINTER);
+    }
+
+    private String formatDate(String year, String month, String day) {
+        return nonNull(year) && !year.isEmpty() ? String.format(FLAT_DATE_TEMPLATE, year, month, day) : null;
```

Comment:
this returns <year>-null-null if month and day are null.

```
private String formatDate(String year, String month, String day) {
        if(isEmpty(year)){
            return null;
        }
        else if (isEmpty(day) && !isEmpty(month)) {
            return formatDateWithoutDay(year, month);
        } else if (isEmpty(month) ) {
            return formatDateWithoutMonth(year);
        } else {
            return formatDateWithYearMonthDay(year, month, day);
        }
    }

    private String formatDateWithYearMonthDay(String year, String month, String day) {
        LocalDate date = LocalDate.of(parseInt(year), parseInt(month), parseInt(day));
        return date.format(DateTimeFormatter.ISO_LOCAL_DATE);
    }

    private boolean isEmpty(String input) {
        return isNull(input) || input.isBlank();
    }

    private String formatDateWithoutDay(String year, String month) {
        LocalDate date = LocalDate.of(parseInt(year), parseInt(month), IGNORED_VALUE);
        return date.format(DateTimeFormatter.ofPattern("yyyy-MM"));
    }

    private String formatDateWithoutMonth(String year) {
        LocalDate date = LocalDate.of(parseInt(year), IGNORED_VALUE, IGNORED_VALUE);
        return date.format(DateTimeFormatter.ofPattern("yyyy"));
    }

private boolean isEmpty(String input) {
        return isNull(input) || input.isBlank();
    }
```

Reply:
Also should we log if the whole date is empty? Or is it expected many times to happen?

Reply:
The expectations are not clear, the date is not obligatory at the moment, we will address this in a subsequent PR